### PR TITLE
`assert`: Report why values aren't equal

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

`assert a == b; x` should report why `a` isn't equal to `b`.

With this change it does so rather well, I would say, especially when `a` and `b` are nested structures. See for example `eval-fail-assert-nested-bool.err.exp`.

# Context

This adds code that duplicates existing behavior. I've done so to keep normal equality fast and simple.
While this approach does add a risk of the code paths diverging in the future, I believe this risk is well worth the benefit.
It makes the difference between knowing immediately why something fails, or a debugging session, changing code in a dependency, etc.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by edolstra** [comment]
Looking at the AST seems pretty hacky and error-prone to me. It would be cleaner to add an `assertEqual` builtin.

**Comment by roberth** [comment]
> Looking at the AST seems pretty hacky and error-prone to me. It would be cleaner to add an `assertEqual` builtin.

Evaluation is just rewriting. Rewrite rule heads matching AST constructors is nice, but not a requirement. I don't see why looking at the AST would be error prone.

I don't like `assertEqual`, because it expands the language unnecessarily, requiring that users learn more and remember more.

I might agree with you in cases where the language semantics is affected, but in this case it is only the error reporting. That's far from critical.

**Review by Mic92 — APPROVED** [review]
I like this approach more than adding a new builtin because we have a lot of existing asserts. This is a nice quality of life improvement like how it currently prints the value when a type miss-matches.

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -2418,6 +2433,216 @@ SingleDerivedPath EvalState::coerceToSingleDerivedPath(const PosIdx pos, Value &
...
+        return;
+
+    case nThunk: // Must not be left by forceValue
+    default:
+        // This should never happen, because eqValues already throws an
+        // error for this, and this function should only be called when
+        // eqValues has found a difference, and it should match
+        // its behavior.
+        error<EvalBaseError>(
+            "cannot compare %1% with %2%; is assertEqValues out of sync with eqValues?", showType(v1), showType(v2))
```
Comment (Mic92) [inline src/libexpr/eval.cc:2640]: Maybe it should say something like `BUG!`. To someone not reading the c++ code ( most of the nix users) it's not clear what "eqValues" mean.

File: src/libexpr/eval.cc [hunk 2]
```diff
@@ -2418,6 +2433,216 @@ SingleDerivedPath EvalState::coerceToSingleDerivedPath(const PosIdx pos, Value &
...
+            "%s of value '%s' is not equal to %s of value '%s'",
+            showType(v1),
+            ValuePrinter(*this, v1, errorPrintOptions),
+            showType(v2),
+            ValuePrinter(*this, v2, errorPrintOptions))
+            .debugThrow();
+    }
+
+    switch (v1.type()) {
+    case nInt:
```
Comment (Mic92) [inline src/libexpr/eval.cc:2475]: Is there a way to do enforce an exhaustive c++ switch case that will fail to compile when there cases missing? Rust has this.

File: src/libexpr/eval.cc (hunk 1 above)
Comment (roberth) [reply src/libexpr/eval.cc]: Changed to:

`"BUG: cannot compare %1% with %2%; did forceValue leave a thunk, or might assertEqValues be out of sync with eqValues?"`

File: src/libexpr/eval.cc (hunk 2 above)
Comment (roberth) [reply src/libexpr/eval.cc:2475]: Good question, but we already enforce this with
```meson
  '-Werror=switch',
  '-Werror=switch-enum',
```

This makes the error message very hypothetical. We'd have to undo those flags, or `forceValue` needs to silently fail, or we'd have to read garbage, or some other chaotic thing we don't need to consider.

Comment (Mic92) [reply src/libexpr/eval.cc:2475]: Could a default case shadow a potential error message and should be therefore removed?

Comment (roberth) [reply src/libexpr/eval.cc:2475]: The flags already prevent this despite the `default`, so the `default` is a bit paranoid.

I've changed the exceptions to crashes to aid debugging, and I've documented that they're unreachable.

File: src/libexpr/eval-error.cc [hunk 3]
```diff
@@ -92,6 +92,14 @@ void EvalErrorBuilder<T>::debugThrow()
     throw error;
 }
 
+template<class T>
+void EvalErrorBuilder<T>::panic()
+{
+    logError(error.info());
+    printError("This is a bug! An unexpected condition occurred, causing the Nix evaluator to have to stop. If you could share a reproducible example or a core dump, please open an issue at https://github.com/NixOS/nix/issues");
+    abort();
```
Comment (roberth) [inline src/libexpr/eval-error.cc:100]: A generalization of this in libutil would be nice for a follow-up.

**Review by edolstra — APPROVED** [review]

**Pull request merged**